research/attention_ocr/python/model.py

```diff
@@ -380,12 +528,29 @@ def create_base(self,
         predicted_text = character_mapper.get_text(predicted_chars)
       else:
         predicted_text = tf.constant([])
+
+      text_log_prob, predicted_length = null_based_length_prediction(
+            chars_log_prob, self._params.null_code)
+      predicted_conf = lookup_indexed_value(predicted_length, text_log_prob)
+      # Convert predicted confidence from sum of logs to geometric mean
```

Comment:
I don't quite understand the reason for this, what am I missing?

Reply:
We want to calculate the probability that the sequence of predicted characters is correct. We assume that the predictions (up to the first null character) are independent (in practice this is an approximation due to the use of the previous character in the prediction of the next one), so the joint probability is the product of probabilities. Since the confidence is the log of the probability, we sum the log probabilities.
The normalized_seq_conf is also provided, so that we can use a single threshold for the sequence confidence that does not depend on the sequence length (longer sequences have lower confidence than shorter ones).

Reply:
> in practice this is an approximation due to the use of the previous character in the prediction of the next one

Thanks for the clarification, I was thinking along these lines but wasn't quite sure. Things make sense now 😄